store/tikv/gcworker/gc_worker.go

```diff
@@ -917,6 +986,313 @@ func (w *GCWorker) resolveLocksForRange(ctx context.Context, safePoint uint64, s
 	return stat, nil
 }
 
+// resolveLocksPhysical uses TiKV's `PhysicalScanLock` to scan stale locks in the cluster and resolve them. It tries to
+// ensure no lock whose ts <= safePoint is left.
+func (w *GCWorker) resolveLocksPhysical(ctx context.Context, safePoint uint64) error {
+	metrics.GCWorkerCounter.WithLabelValues("resolve_locks_physical").Inc()
+	logutil.Logger(ctx).Info("[gc worker] start resolve locks with physical scan locks",
+		zap.String("uuid", w.uuid),
+		zap.Uint64("safePoint", safePoint))
+	startTime := time.Now()
+
+	stores, err := w.getUpStoresMap(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// TODO: If failed anywhere, should we try to cleanup the observers?
+
+	err = w.registerLockObservers(ctx, safePoint, stores)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	for retry := 0; retry < 3; retry++ {
+		resolvedStores, err := w.physicalScanAndResolveLocks(ctx, safePoint, stores)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		stores, err = w.getUpStoresMap(ctx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		checkedStores, err := w.checkLockObservers(ctx, safePoint, stores)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		// Remove clean stores from the set
+		for resolvedStore := range resolvedStores {
+			// Only stores that are both resolved and checked is clean.
+			// For each clean store, remove it from the stores set.
+			if _, ok := checkedStores[resolvedStore]; ok {
+				delete(stores, resolvedStore)
+			}
+		}
+
+		// If there are still dirty stores, continue the loop to clean them again.
+		// Only dirty stores will be scanned in the next loop.
+		if len(stores) == 0 {
+			break
+		}
+	}
+
+	err = w.removeLockObservers(ctx, safePoint, stores)
+	if err != nil {
+		return errors.Trace(err)
```

Comment:
If remove observer fails, we can still omit the traditional way?

Reply:
Sounds good

Reply:
It already is because removeLockObserver never returns error... I've removed the return value